File: tugas_LL/src/CircularLL/circular_doublyLL_berita.py
import time
import sys
import threading
import shutil


SCROLL_SPEED = 0.2   # kecepatan animasi huruf
DELAY_BETWEEN = 3     # delay antar berita (3 detik)

#WIDTH = flexible, disesuaikan dengan lebar terminal
def get_terminal_width():
    return shutil.get_terminal_size().columns

def wait_for_enter(stop_event):
		input()  # tunggu ENTER
		stop_event.set()

# Node class for CDLL
class Node:
	def __init__(self, berita):
		self.berita = berita
		self.next = None
		self.prev = None


# Circular Doubly Linked List class
class CircularDoublyLinkedList:
	def __init__(self):
		self.head = None
		self.size = 0
	
	

	# Insert berita di akhir list
	def insert(self, berita):
		"""Menambahkan berita di akhir list."""
		new_node = Node(berita)
		if self.head is None:
			self.head = new_node
			new_node.next = new_node
			new_node.prev = new_node
		else:
			tail = self.head.prev
			tail.next = new_node
			new_node.prev = tail
			new_node.next = self.head
			self.head.prev = new_node
		self.size += 1

	# Hapus berita berdasarkan nomor urut (index mulai dari 1)
	def delete_by_index(self, index):
		"""Menghapus berita berdasarkan nomor urut (mulai dari 1)."""
		if self.head is None or index < 1 or index > self.size:
			return False
		if self.size == 1:
			self.head = None
			self.size = 0
			return True
		current = self.head
		if index == 1:
			tail = self.head.prev
			self.head = self.head.next
			tail.next = self.head
			self.head.prev = tail
			self.size -= 1
			return True
		for _ in range(1, index):
			current = current.next
		prev_node = current.prev
		next_node = current.next
		prev_node.next = next_node
		next_node.prev = prev_node
		self.size -= 1
		return True
	
	

	# Tampilkan berita forward dengan delay 3 detik
	def display_forward(self):
		if self.head is None:
			print("Belum ada berita.")
			return

		stop_event = threading.Event()
		threading.Thread(target=wait_for_enter, args=(stop_event,), daemon=True).start()

		print("Tekan ENTER untuk kembali ke menu...\n")

		current = self.head
		no = 1

		while not stop_event.is_set():

			text = f"[{no}] {current.berita}"
			width = shutil.get_terminal_size().columns
			padding = " " * width
			scroll_text = padding + text + padding

			position = 0

			# Animasi kanan ➜ kiri
			while position + width <= len(scroll_text) and not stop_event.is_set():
				frame = scroll_text[position:position + width]
				sys.stdout.write("\r" + frame)
				sys.stdout.flush()
				time.sleep(SCROLL_SPEED)
				position += 1

			# Delay 3 detik antar berita (tetap bisa keluar saat delay)
			start = time.time()
			while time.time() - start < DELAY_BETWEEN:
				if stop_event.is_set():
					break
				time.sleep(0.1)

			current = current.next
			if current == self.head:
				no = 1
			else:
				no += 1

		# Bersihkan baris
		width = get_terminal_width()
		sys.stdout.write("\r" + " " * width + "\r")
		print("Kembali ke menu...")

	# Tampilkan berita backward dengan delay 3 detik
	def display_backward(self):
		if self.head is None:
			print("Belum ada berita.")
			return

		stop_event = threading.Event()
		threading.Thread(target=wait_for_enter, args=(stop_event,), daemon=True).start()

		print("Tekan ENTER untuk kembali ke menu...\n")

		current = self.head.prev
		no = self.size

		while not stop_event.is_set():

			text = f"[{no}] {current.berita}"
			width = shutil.get_terminal_size().columns
			padding = " " * width
			scroll_text = padding + text + padding

			position = 0

			# Animasi kanan ➜ kiri
			while position + width <= len(scroll_text) and not stop_event.is_set():
				frame = scroll_text[position:position + width]
				sys.stdout.write("\r" + frame)
				sys.stdout.flush()
				time.sleep(SCROLL_SPEED)
				position += 1

			# Delay 3 detik antar berita
			start = time.time()
			while time.time() - start < DELAY_BETWEEN:
				if stop_event.is_set():
					break
				time.sleep(0.1)

			current = current.prev
			if current == self.head.prev:
				no = self.size
			else:
				no -= 1

		width = get_terminal_width()
		sys.stdout.write("\r" + " " * width + "\r")
		print("Kembali ke menu...")

	# Tampilkan berita tertentu berdasarkan nomor urut
	def display_by_index(self, index):
		"""Menampilkan berita tertentu berdasarkan nomor urut."""
		if self.head is None or index < 1 or index > self.size:
			print("Nomor berita tidak valid.")
			return
		current = self.head
		for _ in range(1, index):
			current = current.next
		print(f"{index}. {current.berita}")

	# Tampilkan semua berita (tanpa delay, untuk debug)
	def display_all(self):
		if self.head is None:
			print("Belum ada berita.")
			return
		current = self.head
		no = 1
		while True:
			print(f"{no}. {current.berita}")
			current = current.next
			no += 1
			if current == self.head:
				break

def clean_text(text):
	"""Membersihkan spasi depan, belakang, dan spasi ganda."""
	if text is None:
		return ''
	text = text.strip()
	text = ' '.join(text.split())
	return text

def main():
	cdll = CircularDoublyLinkedList()
	while True:
		print("===== MENU BERITA =====")
		print("1. Insert berita")
		print("2. Hapus berita")
		print("3. Tampilkan berita Forward (delay 3 detik)")
		print("4. Tampilkan berita Backward (delay 3 detik)")
		print("5. Tampil berita tertentu")
		print("6. Exit")
		pilihan = input("Pilih menu: ").strip()
		if not pilihan:
			continue
		if pilihan == '1':
			berita = input("Masukkan berita: ")
			if not berita:
				continue
			berita = clean_text(berita)
			if not berita:
				print("Error: Berita tidak boleh kosong!")
				continue
			cdll.insert(berita)
			print("Berita berhasil ditambahkan.")
		elif pilihan == '2':
			idx_str = input("Masukkan nomor urut berita yang akan dihapus: ").strip()
			if not idx_str:
				continue
			if not idx_str.isdigit():
				print("Error: Input harus berupa angka!")
				continue
			idx = int(idx_str)
			if not cdll.delete_by_index(idx):
				print("Error: Nomor urut tidak valid!")
			else:
				print("Berita berhasil dihapus.")
		elif pilihan == '3':
			cdll.display_forward()
		elif pilihan == '4':
			cdll.display_backward()
		elif pilihan == '5':
			idx_str = input("Masukkan nomor urut berita: ").strip()
			if not idx_str:
				continue
			if not idx_str.isdigit():
				print("Error: Input harus berupa angka!")
				continue
			idx = int(idx_str)
			cdll.display_by_index(idx)
		elif pilihan == '6':
			print("Terima kasih!")
			break
		elif pilihan == '7':
			cdll.display_all() 	# menu debug
		else:
			print("Menu tidak valid!")
		print()

if __name__ == "__main__":
	main()

File: tugas_LL/src/CircularLL/Circular_DoublyLL_Berita.java
package CircularLL;

import java.util.Scanner;

// Node class for CDLL
class Node {
	String berita;
	Node next;
	Node prev;

	Node(String berita) {
		this.berita = berita;
		this.next = null;
		this.prev = null;
	}
}

// Circular Doubly Linked List class
class CircularDoublyLinkedList {
	Node head;
	int size;

	public CircularDoublyLinkedList() {
		head = null;
		size = 0;
	}

	// Insert berita di akhir list
	public void insert(String berita) {
		Node newNode = new Node(berita);
		if (head == null) {
			head = newNode;
			head.next = head;
			head.prev = head;
		} else {
			Node tail = head.prev;
			tail.next = newNode;
			newNode.prev = tail;
			newNode.next = head;
			head.prev = newNode;
		}
		size++;
	}

	// Hapus berita berdasarkan nomor urut (index mulai dari 1)
	public boolean deleteByIndex(int index) {
		if (head == null || index < 1 || index > size) {
			return false;
		}
		if (size == 1) {
			head = null;
			size = 0;
			return true;
		}
		Node current = head;
		if (index == 1) {
			Node tail = head.prev;
			head = head.next;
			tail.next = head;
			head.prev = tail;
			size--;
			return true;
		}
		for (int i = 1; i < index; i++) {
			current = current.next;
		}
		Node prevNode = current.prev;
		Node nextNode = current.next;
		prevNode.next = nextNode;
		nextNode.prev = prevNode;
		// If deleting tail
		if (current == head.prev) {
			// nothing extra needed, already handled
		}
		size--;
		return true;
	}

	// Tampilkan berita forward dengan delay 3 detik
	public void displayForward() {
		if (head == null) {
			System.out.println("Belum ada berita.");
			return;
		}
		Node current = head;
		int no = 1;
		do {
			System.out.println(no + ". " + current.berita);
			try {
				Thread.sleep(3000);
			} catch (InterruptedException e) {
				Thread.currentThread().interrupt();
			}
			current = current.next;
			no++;
		} while (current != head);
	}

	// Tampilkan berita backward dengan delay 3 detik
	public void displayBackward() {
		if (head == null) {
			System.out.println("Belum ada berita.");
			return;
		}
		Node tail = head.prev;
		Node current = tail;
		int no = size;
		do {
			System.out.println(no + ". " + current.berita);
			try {
				Thread.sleep(3000);
			} catch (InterruptedException e) {
				Thread.currentThread().interrupt();
			}
			current = current.prev;
			no--;
		} while (current != tail);
	}

	// Tampilkan berita tertentu berdasarkan nomor urut
	public void displayByIndex(int index) {
		if (head == null || index < 1 || index > size) {
			System.out.println("Nomor berita tidak valid.");
			return;
		}
		Node current = head;
		for (int i = 1; i < index; i++) {
			current = current.next;
		}
		System.out.println(index + ". " + current.berita);
	}

	// Tampilkan semua berita (tanpa delay, untuk debug)
	public void displayAll() {
		if (head == null) {
			System.out.println("Belum ada berita.");
			return;
		}
		Node current = head;
		int no = 1;
		do {
			System.out.println(no + ". " + current.berita);
			current = current.next;
			no++;
		} while (current != head);
	}
}

public class Circular_DoublyLL_Berita {
	// Membersihkan spasi depan, belakang, dan spasi ganda
	public static String cleanText(String text) {
		if (text == null) return "";
		text = text.trim();
		text = text.replaceAll("\\s+", " ");
		return text;
	}

	public static void main(String[] args) {
		try (Scanner sc = new Scanner(System.in)) {
            CircularDoublyLinkedList cdll = new CircularDoublyLinkedList();
            while (true) {
                System.out.println("===== MENU BERITA =====");
                System.out.println("1. Insert berita (ditambahkan di akhir)");
                System.out.println("2. Hapus berita (berdasarkan nomor urut)");
                System.out.println("3. Tampilkan berita Forward (delay 3 detik)");
                System.out.println("4. Tampilkan berita Backward (delay 3 detik)");
                System.out.println("5. Tampil berita tertentu (berdasarkan nomor urut)");
                System.out.println("6. Exit");
                System.out.print("Pilih menu: ");

                String input = sc.nextLine();
                if (input.isEmpty()) continue;

                switch (input) {
                    case "1":
                        System.out.print("Masukkan berita: ");
                        String berita = sc.nextLine();
                        if (berita.isEmpty()) break;
                        berita = cleanText(berita);
                        if (berita.isEmpty()) {
                            System.out.println("Error: Berita tidak boleh kosong!");
                            break;
                        }
                        cdll.insert(berita);
                        System.out.println("Berita berhasil ditambahkan.");
                        break;
                    case "2":
                        System.out.print("Masukkan nomor urut berita yang akan dihapus: ");
                        String idxStr = sc.nextLine();
                        if (idxStr.isEmpty()) break;
                        if (!idxStr.matches("\\d+")) {
                            System.out.println("Error: Input harus berupa angka!");
                            break;
                        }
                        int idx = Integer.parseInt(idxStr);
                        if (!cdll.deleteByIndex(idx)) {
                            System.out.println("Error: Nomor urut tidak valid!");
                        } else {
                            System.out.println("Berita berhasil dihapus.");
                        }
                        break;
                    case "3":
                        cdll.displayForward();
                        break;
                    case "4":
                        cdll.displayBackward();
                        break;
                    case "5":
                        System.out.print("Masukkan nomor urut berita: ");
                        String idxStr2 = sc.nextLine();
                        if (idxStr2.isEmpty()) break;
                        if (!idxStr2.matches("\\d+")) {
                            System.out.println("Error: Input harus berupa angka!");
                            break;
                        }
                        int idx2 = Integer.parseInt(idxStr2);
                        cdll.displayByIndex(idx2);
                        break;
                    case "6":
                        System.out.println("Terima kasih!");
                        return;
                    case "7":
                        cdll.displayAll();
                        break;
                    default:
                        System.out.println("Menu tidak valid!");
                        
                }
                
                System.out.println();
            }
		}
	}
}
